lib/codecs/src/encoding/format/protobuf.rs

```diff
@@ -0,0 +1,342 @@
+        assert!(mfield!(message, "f").as_f32() == Some(2.0));
+    }
+
+    #[test]
+    fn test_encode_bytes() {
+        let bytes = Bytes::from(vec![0, 1, 2, 3]);
+        let message = encode_message(
+            &test_message_descriptor("Bytes"),
+            Value::Object(BTreeMap::from([
+                ("text".into(), Value::Bytes(Bytes::from("vector"))),
+                ("binary".into(), Value::Bytes(bytes.clone())),
+            ])),
+        )
+        .unwrap();
+        assert!(mfield!(message, "text").as_str() == Some("vector"));
+        assert!(mfield!(message, "binary").as_bytes() == Some(&bytes));
+    }
+
+    #[test]
+    fn test_encode_repeated_primitive() {
+        let message = encode_message(
+            &test_message_descriptor("RepeatedPrimitive"),
+            Value::Object(BTreeMap::from([(
+                "numbers".into(),
+                Value::Array(vec![
+                    Value::Integer(8),
+                    Value::Integer(6),
+                    Value::Integer(4),
+                ]),
+            )])),
+        )
+        .unwrap();
+        let list = mfield!(message, "numbers").as_list().unwrap().to_vec();
+        assert!(list.len() == 3);
+        assert!(list[0].as_i64() == Some(8));
+        assert!(list[1].as_i64() == Some(6));
+        assert!(list[2].as_i64() == Some(4));
+    }
+
+    #[test]
+    fn test_encode_repeated_message() {
+        let message = encode_message(
+            &test_message_descriptor("RepeatedMessage"),
+            Value::Object(BTreeMap::from([(
+                "messages".into(),
+                Value::Array(vec![
+                    Value::Object(BTreeMap::from([(
+                        "text".into(),
+                        Value::Bytes(Bytes::from("vector")),
+                    )])),
+                    Value::Object(BTreeMap::from([("index".into(), Value::Integer(4444))])),
+                    Value::Object(BTreeMap::from([
+                        ("text".into(), Value::Bytes(Bytes::from("protobuf"))),
+                        ("index".into(), Value::Integer(1)),
+                    ])),
+                ]),
+            )])),
+        )
+        .unwrap();
+        let list = mfield!(message, "messages").as_list().unwrap().to_vec();
+        assert!(list.len() == 3);
+        assert!(mfield!(list[0].as_message().unwrap(), "text").as_str() == Some("vector"));
+        assert!(!list[0].as_message().unwrap().has_field_by_name("index"));
+        assert!(!list[1].as_message().unwrap().has_field_by_name("t4ext"));
+        assert!(mfield!(list[1].as_message().unwrap(), "index").as_u32() == Some(4444));
+        assert!(mfield!(list[2].as_message().unwrap(), "text").as_str() == Some("protobuf"));
+        assert!(mfield!(list[2].as_message().unwrap(), "index").as_u32() == Some(1));
+    }
+
+    #[test]
+    fn test_encode_decoding_protobuf_test_data() {
+        let test_data_dir = PathBuf::from(std::env::var_os("CARGO_MANIFEST_DIR").unwrap())
```

Comment:
💡 We could put all these under `tests/data/protobuf` since we are using them here as well.

Reply:
Moved all test constructs to `tests/data/protobuf`.